### ISSUE TITLE
No narrowing of related TypeVars

### ISSUE THREAD

**Issue by jonasnylund** [microsoft/pyright#11097]
I have run across this issue in pyright and concluded it must be a bug. When calling an overloaded function from another with TypeVar arguments, pyright seem unable to narrow the types of the TypeVar correctly.

```python
from typing import overload, TypeVar

_T = TypeVar("_T", int, str)

@overload
def add1(a: int, b: int) -> int: ...
@overload
def add1(a: str, b: str) -> str: ...

def add1(a, b):
    return a + b

def add2(a: _T, b: _T) -> _T:
    return add1(a, b)
```
Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQAqixAapSAFAcD6DUAvI2ZsQACgBEvcfVQx6AZxggAlFwACRUhRodqxYFBrUAjKMoAuTCjlQARpdnKoAWgB8VmJYB0PjhpLkVNS6%2BobUJmaWiiD09lDRTm7xSt6%2BIQZGppSxyuYcUAVQIMQwAK4gKIZQANR2XHoZ4QBMkVC8sZa8ie68eYVFJeWVmWY5HEA)

This gives the following error on the last line

> No overloads for "add1" match the provided arguments
  Argument of type "_T@add2" cannot be assigned to parameter "a" of type "str" in function "add1"
    Type "int* | str*" is not assignable to type "str"
      "int*" is not assignable to "str"

According to the last defining example in the overload section in PEP 484 https://peps.python.org/pep-0484/#function-method-overloading, the two implementations of `addx` in the code example above should be equivalent from a type checking POV, which in my mind sounds like I should be able to do this.

I also find the following example strange

```python
from typing import TypeVar, reveal_type

_T = TypeVar("_T")

def add2(a: _T, b: _T):
    if isinstance(a, int):
        reveal_type(b)  # Type of "b" is "_T@add2"
```

I would expect `b` to be narrowed in the scope of the if-statement to the same type as `a`, but no narrowing has taken place. It is almost as if pyright is treating the TypeVar as a Type Union in this case. I would assume this issue to be related to the one above. This also seems inconsistent with the documented behavior in https://microsoft.github.io/pyright/#/type-concepts-advanced?id=value-constrained-type-variables

I am surprised to find no other issues relating to this on the github open issues, it seems like such a basic case that should have been noticed already?

These issues are observed in pyright 1.1.407

**Comment by erictraut** [comment]
You're using a value-constrained type variable here. The behaviors for these are underspecified in the Python typing spec. It's not clear to me that we'll ever be able to specify and reconcile their behaviors with other aspects of the typing spec. It's unfortunate that this feature was introduced to Python. I recommend avoiding their use. There's almost always a better alternative. You won't see the concept of value-constrained type variables in other typed languages, and there's a good reason for that.

As for type narrowing, it applies to individual expressions. When one expression is narrowed, it doesn't affect the narrowed type of another expression. In your example above, the type of `a` is narrowed based on the type guard `isinstance(a, int)`, but this doesn't affect the narrowed type of the expression `b`. For more details about how type narrowing works in pyright, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing).

In short, pyright is working as designed here.
